app/models/assignment.rb

```diff
@@ -851,7 +853,72 @@ def get_num_valid
              .count(&:is_valid?)
   end
 
-  def get_num_marked(ta_id = nil)
+  def marked_result_ids_for(ta_id)
+    cache_ta_results.dig(ta_id, :marked_result_ids) || []
+  end
+
+  def cache_ta_results
+    return @cache_ta_results if defined? @cache_ta_results
+
+    data = current_results.joins(grouping: :tas).pluck('tas.id', 'results.id', 'results.marking_state')
+    # Group results by TA ID
+    grouped_data = data.group_by { |ta_id, _result_id, _marking_state| ta_id }
+
+    results_by_id = current_results.includes(:marks).index_by(&:id)
+
+    ta_to_criteria = self.criterion_ta_associations.group_by(&:ta_id).transform_values do |assocs|
+      assocs.map(&:criterion_id)
+    end
+
+    @cache_ta_results = grouped_data.map do |ta_id, results|
+      total_results = results.map { |_, result_id, _| result_id }
+
+      if self.assign_graders_to_criteria
+        # Get the list of criteria IDs assigned to this TA
+        assigned_criteria_ids = ta_to_criteria[ta_id] || []
+        criteria_count = assigned_criteria_ids.size
+
+        if criteria_count == 0
+          # If the TA has no assigned criteria, fallback to using marking_state to determine marked results
+          complete_results = results.select do |_, _, marking_state|
+            marking_state ==
+                      Result::MARKING_STATES[:complete]
+          end
+          marked_result_ids = complete_results.map { |_, result_id, _| result_id }
+        else
+          # Count results where all assigned criteria have been marked
+          total_results.select do |result_id|
+            result = results_by_id[result_id]
+            next false unless result
+
+            marks_for_ta = result.marks.select do |mark|
+              assigned_criteria_ids.include?(mark.criterion_id) && !mark.mark.nil?
+            end
+            # Only include this result if all criteria assigned to the TA have been marked
+            marks_for_ta.size == criteria_count
+          end
+          marked_result_ids = Result.joins(:marks)
+                                    .where(id: total_results, marks: { criterion_id: assigned_criteria_ids })
+                                    .where.not(marks: { mark: nil })
+                                    .group('results.id')
+                                    .having('count(marks.id) = ?', criteria_count)
+                                    .pluck('results.id')
+        end
+      else
+        # Grading not by criterion: count only completed results
+        marked_result_ids = results.select { |_, _, marking_state| marking_state == Result::MARKING_STATES[:complete] }
+                                   .map { |_, result_id, _| result_id }
+      end
+      # Save the results into the cache
```

Comment:
I don't think this comment really applies any more, and you can safely remove it